### ISSUE TITLE
False error when values of dictionary (which are functions) are assigned to variable

### ISSUE THREAD

**Issue by smallpondtom** [microsoft/pyright#3248]
**Describe the bug**
I am getting a false error when trying to assign a dictionary value (which is function) to a variable inside a class. The error message is
```
Argument of type "str | FunctionType" cannot be assigned to parameter "__k" of type "str" in function "__getitem__"  Type "str | FunctionType" cannot be assigned to type "str"    "FunctionType" is incompatible with "str" Pyright (reportGeneralTypeIssues) [55, 27]
```
The exact location are the lines
```
 self.solver = METHOD[method]    
 self.p = ORDER[method]    
``` 
in the code snippet below.

**To Reproduce**
conda environment with
- python = 3.10
- numpy
pyright -> latest version

**Expected behavior**
There should be no errors.

**Screenshots or Code**
```python
import inspect                                                           
import numpy as np                                                       
from numpy.typing import NDArray                                         
from typing import Callable, Optional, Any, Union                        
                                                                         
                                                                         
ODE = Union[                                                             
    Callable[[float, NDArray[Any], Optional[Any]],                       
             Union[NDArray[Any], NDArray[np.float64]]],                  
    Callable[[float, NDArray[Any]],                                      
             Union[NDArray[Any], NDArray[np.float64]]]
]
def RKF45():
    return 0

def RKF78():
    return 1



METHOD = {
    "RKF45": RKF45,
    "RKF78": RKF78
}

ORDER = {
    "CKRK45": 4,
    "DOP54": 5,
    "DVERK65": 6,
    "Euler": 1,
    "RALS4": 4,
    "RK4": 4,
    "RK42": 4,
    "RKF45": 4,
    "RKF78": 7
}

class check:

    def __init__(self, derivative: ODE, h: float, tspan: NDArray[Any],
                 IC: NDArray[np.float64], Atol: float, Rtol: float,
                 method: str = "RKF78"):
        self.ode = derivative
        self.h = h
        self.tspan = tspan
        self.IC = IC
        self.Atol = Atol
        self.Rtol = Rtol


        # Check solver method and stepsize controller
        if method not in METHOD and not inspect.isfunction(method):
            raise ValueError("`method` must be one of {}".format(METHOD))
        else:
            self.solver = METHOD[method]    
            self.p = ORDER[method]    


# Test ------------------------------------------------------------------------
def test(t: float, y: NDArray[np.float64]) -> NDArray[Any]:
    return np.array([
        y[1],
        -4*t*y[1] - (2 + 4*t**2)*y[0]
    ])


if __name__=="__main__":

    # Setup values
    tf = 20  # final time
    t = 0   # initial time
    ti = t
    x = np.array([1, 0.5])  # initial conditions of states
    h = 0.01  # initial stepsize
    p = 4  # order of the numerical integration
    atol = 1e-7  # absolute tolerance
    rtol = 1e-6  # relative tolerance

    ck = check(test, h, np.array([ti, tf]), x, atol, rtol) 
    print(ck.solver)
    print(ck.p)             
```

**VS Code extension or command-line**
Running on command-line. Neovim with Native LSP on Pyright = 1.1.232

**Comment by erictraut** [comment]
I think the current behavior is correct.

You have declared parameter `method` as being type `str`, but you later apply the check `inspect.isfunction(method)`. This function is annotated to be a "type guard", and if the check succeeds, it indicates that the expression should be "narrowed" to the type FunctionType.

Under what circumstance will `method` ever be a function? That doesn't look possible given that you have annotated the parameter with the type `str`. Any attempt to pass a function (an actual callable object) will result in type checking errors. 

Also, if `method` is a function, you will attempt to index into the dict `METHOD` with a function as a key, but a function is not a hashable object, so this will fail.

So, if I understand your code correctly, you should delete the code ` and not inspect.isfunction(method)`.  This will eliminate the type error.
